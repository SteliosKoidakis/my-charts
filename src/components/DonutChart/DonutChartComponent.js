import {
  formatNumber,
  renderStyles,
  renderDonutChart,
  renderLineChart,
} from '~/utils';
import DonutChartComponentStyles from './DonutChartComponent.scss';

const DONUT_CHART_CLASS = 'donutChart';
const LINE_CHART_CLASS = 'lineChart';
const COMPONENT_CLASS = 'DonutChartComponent';

class DonutChartComponent extends HTMLElement {
  constructor() {
    super();
    this.shadow = this.attachShadow({
      mode: 'open',
    });
  }

  static get observedAttributes() {
    return [
      'currency',
      'smartphonePercentage',
      'tabletPercentage',
      'smartphone',
      'tablet',
      'smartphoneColor',
      'tabletColor',
      'total',
      'title',
      'data',
    ];
  }

  get tabletPercentage() {
    return parseFloat(this.getAttribute('tabletPercentage')) || 0;
  }

  get smartphonePercentage() {
    return parseFloat(this.getAttribute('smartphonePercentage')) || 0;
  }

  get tablet() {
    return parseFloat(this.getAttribute('tablet')) || 0;
  }

  get smartphone() {
    return parseFloat(this.getAttribute('smartphone')) || 0;
  }

  get tabletColor() {
    return this.getAttribute('tabletColor');
  }

  get currency() {
    return this.getAttribute('currency') || '';
  }

  get smartphoneColor() {
    return this.getAttribute('smartphoneColor') || ' ';
  }

  get title() {
    return this.getAttribute('title') || '';
  }

  get total() {
    return parseFloat(this.getAttribute('total')) || 0;
  }

  get data() {
    return this.getAttribute('data') && this.getAttribute('data').split(',');
  }

  async connectedCallback() {
    this.renderTemplate();
    renderStyles(this.shadow, DonutChartComponentStyles);
    renderDonutChart({
      smartphone: {
        value: this.smartphone,
        smartphoneColor: this.smartphoneColor,
      },
      tablet: {
        value: this.tablet,
        tabletColor: this.tabletColor,
      },
      element: {
        shadow: this.shadow,
        class: DONUT_CHART_CLASS,
      },
    });
    renderLineChart({
      data: this.data,
      color: this.smartphoneColor,
      element: {
        shadow: this.shadow,
        class: LINE_CHART_CLASS,
        backgroundClass: `${COMPONENT_CLASS}__background__${this.title.toLowerCase()}`,
      },
    });
  }

  renderTemplate() {
    const template = `
      <div class="${COMPONENT_CLASS} ${COMPONENT_CLASS}--${this.title.toLowerCase()}">
        <div class="${COMPONENT_CLASS}__donut ${DONUT_CHART_CLASS}">
          <div class="${COMPONENT_CLASS}__line-chart ${LINE_CHART_CLASS}">
            <strong class="${COMPONENT_CLASS}__title">${this.title}</strong>
            <strong class="${COMPONENT_CLASS}__total">${formatNumber(this.total)}</strong>
          </div>
        </div>
        <div class="${COMPONENT_CLASS}__info">
          <div class="${COMPONENT_CLASS}__tablet-info">
            <strong class="${COMPONENT_CLASS}__tablet-text">Tablet</strong>
            <span>${this.tabletPercentage}%</span>
            <span class="${COMPONENT_CLASS}__value">${formatNumber(this.tablet)}${this.currency}</span>
          </div>
          <div class="${COMPONENT_CLASS}__smartphone-info">
            <strong class="${COMPONENT_CLASS}__smartphone-text">Smartphone</strong>
            <span>${this.smartphonePercentage}%</span>
            <span class="${COMPONENT_CLASS}__value">${formatNumber(this.smartphone)}${this.currency}</span>
          </div>
        </div>
      </div>
    `;

    this.shadow.innerHTML = template;
  }
}

window.customElements.define('donut-chart-component', DonutChartComponent);

export default DonutChartComponent;
