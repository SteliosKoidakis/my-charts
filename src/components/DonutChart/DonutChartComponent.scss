.DonutChartComponent {
  $blockClass: &;
  position: relative;
  
  &__donut {
    position: relative;
    padding: 0 40px;
  }

  &__title {
    color: grey;
    font-size: 15px;
  }

  &__total {
    font-size: 18px;
  }

  &__title,
  &__total {
    display: block;
    padding-bottom: 5px;
    text-align: center;
  }

  &__line-chart {
    position: absolute;
    bottom: 40px;
    left: 80px;
  }

  &__info {
    display: flex;
    justify-content: space-between;
  }

  &__tablet-info {
    text-align: left;
  }

  &__smartphone-info {
    text-align: right;
  }

  &__value {
    color: lightgrey;
  }

  &__background {

    &__revenue {
      fill: #e4fcde;
    }
  
    &__impressions {
      fill: #dbfdff;
    }
  
    &__visits {
      fill: #fff0ed;
    }
  }

  &--revenue {
  
    #{$blockClass}__tablet-text {
      color: #44DF22;
    }
  
    #{$blockClass}__smartphone-text {
      color: #38AF1E;
    }
  }

  &--impressions {
  
    #{$blockClass}__tablet-text {
      color: #4DD4DD;
    }

    #{$blockClass}__smartphone-text {
      color: #2945BB;
    }
  }

  &--visits {

    #{$blockClass}__tablet-text {
      color: #E5C839;
    }

    #{$blockClass}__smartphone-text {
      color: #E55039;
    }
  }

  &__tablet-text,
  &__smartphone-text {
    display: block;
  }
}