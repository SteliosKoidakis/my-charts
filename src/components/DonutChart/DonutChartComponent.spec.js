import DonutChartComponent from './DonutChartComponent';
import {
  renderStyles,
  renderDonutChart,
  renderLinarChart,
} from '~/utils';

jest.mock('~/utils', () => ({
  formatNumber: jest.fn(),
  renderStyles: jest.fn(),
  renderDonutChart: jest.fn(),
  renderLinarChart: jest.fn(),
}));

describe('Given component DonutChartComponent', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = new DonutChartComponent();
    jest.resetAllMocks();
  });
  afterEach(() => {
    jest.restoreAllMocks();
  });
  describe('Given tabletPercentage getter', () => {
    describe('When tabletPercentage attribute is a falsy or unexpected value', () => {
      it('Then it should return 0', () => {
        wrapper.setAttribute('tabletPercentage', false);
        expect(wrapper.tabletPercentage).toBe(0);
        wrapper.setAttribute('tabletPercentage', {});
        expect(wrapper.tabletPercentage).toBe(0);
      });
    });
    describe('When tabletPercentage attribute is a number in a string format', () => {
      it('Then it should return the number', () => {
        wrapper.setAttribute('tabletPercentage', '2');
        expect(wrapper.tabletPercentage).toBe(2);
      });
    });
  });
  describe('Given smartphonePercentage getter', () => {
    describe('When smartphonePercentage attribute is a falsy or unexpected value', () => {
      it('Then it should return 0', () => {
        wrapper.setAttribute('smartphonePercentage', false);
        expect(wrapper.smartphonePercentage).toBe(0);
        wrapper.setAttribute('smartphonePercentage', {});
        expect(wrapper.smartphonePercentage).toBe(0);
      });
    });
    describe('When smartphonePercentage attribute is a number in a string format', () => {
      it('Then it should return the number', () => {
        wrapper.setAttribute('smartphonePercentage', '2');
        expect(wrapper.smartphonePercentage).toBe(2);
      });
    });
  });
  describe('Given tablet getter', () => {
    describe('When tablet attribute is a falsy or unexpected value', () => {
      it('Then it should return 0', () => {
        wrapper.setAttribute('tablet', false);
        expect(wrapper.tablet).toBe(0);
        wrapper.setAttribute('tablet', {});
        expect(wrapper.tablet).toBe(0);
      });
    });
    describe('When tablet attribute is a number in a string format', () => {
      it('Then it should return the number', () => {
        wrapper.setAttribute('tablet', '2');
        expect(wrapper.tablet).toBe(2);
      });
    });
  });
  describe('Given smartphone getter', () => {
    describe('When smartphone attribute is a falsy or unexpected value', () => {
      it('Then it should return 0', () => {
        wrapper.setAttribute('smartphone', false);
        expect(wrapper.smartphone).toBe(0);
        wrapper.setAttribute('smartphone', {});
        expect(wrapper.smartphone).toBe(0);
      });
    });
    describe('When smartphone attribute is a number in a string format', () => {
      it('Then it should return the number', () => {
        wrapper.setAttribute('smartphone', '2');
        expect(wrapper.smartphone).toBe(2);
      });
    });
  });
  describe('Given tabletColor getter', () => {
    describe('When tabletColor attribute has any value', () => {
      it('Then it should return it in string format', () => {
        wrapper.setAttribute('tabletColor', false);
        expect(typeof wrapper.tabletColor).toBe('string');
        wrapper.setAttribute('tabletColor', {});
        expect(typeof wrapper.tabletColor).toBe('string');
      });
    });
  });
  describe('Given currency getter', () => {
    describe('When currency attribute has any value', () => {
      it('Then it should return ""', () => {
        wrapper.setAttribute('currency', false);
        expect(typeof wrapper.currency).toBe('string');
        wrapper.setAttribute('currency', {});
        expect(typeof wrapper.currency).toBe('string');
      });
    });
  });
  describe('Given smartphoneColor getter', () => {
    describe('When smartphoneColor attribute has any value', () => {
      it('Then it should return it in a string format', () => {
        wrapper.setAttribute('smartphoneColor', false);
        expect(typeof wrapper.smartphoneColor).toBe('string');
        wrapper.setAttribute('smartphoneColor', {});
        expect(typeof wrapper.smartphoneColor).toBe('string');
      });
    });
  });
  describe('Given title getter', () => {
    describe('When title attribute has any value', () => {
      it('Then it should return it in a string format', () => {
        wrapper.setAttribute('title', false);
        expect(typeof wrapper.title).toBe('string');
        wrapper.setAttribute('title', {});
        expect(typeof wrapper.title).toBe('string');
      });
    });
  });
  describe('Given data getter', () => {
    describe('When data attribute has any value', () => {
      it('Then it should return an array with values split by comma', () => {
        wrapper.setAttribute('data', false);
        expect(wrapper.data).toStrictEqual(['false']);
      });
    });
  });
  describe('Given total getter', () => {
    describe('When total attribute is a falsy or unexpected value', () => {
      it('Then it should return 0', () => {
        wrapper.setAttribute('total', false);
        expect(wrapper.total).toBe(0);
        wrapper.setAttribute('total', {});
        expect(wrapper.total).toBe(0);
      });
    });
    describe('When total attribute is a number in a string format', () => {
      it('Then it should return the number', () => {
        wrapper.setAttribute('total', '2');
        expect(wrapper.total).toBe(2);
      });
    });
  });
  describe('Given connectedCallback method', () => {
    describe('When is called', () => {
      it('Then it should trigger renderTemplate and renderStyles method', async () => {
        wrapper.renderTemplate = jest.fn();
        wrapper.renderStyles = jest.fn();
        await wrapper.connectedCallback();
        expect(wrapper.renderTemplate).toHaveBeenCalled();
        expect(renderStyles).toHaveBeenCalled();
      });
      it('And it should trigger renderDonutChart with the correct parameter', async () => {
        wrapper.setAttribute('smartphone', 2);
        wrapper.setAttribute('smartphoneColor', 'smartphoneColor');
        wrapper.setAttribute('tablet', 2);
        wrapper.setAttribute('tabletColor', 'tabletColor');
        await wrapper.connectedCallback();
        expect(renderDonutChart).toHaveBeenCalledWith({
          smartphone: {
            value: 2,
            smartphoneColor: 'smartphoneColor',
          },
          tablet: {
            value: 2,
            tabletColor: 'tabletColor',
          },
          element: {
            shadow: wrapper.shadow,
            class: 'donutChart',
          },
        });
      });
      it('And it should trigger renderLinarChart with the correct parameter', async () => {
        const data = '1,2,3';
        wrapper.setAttribute('data', data);
        wrapper.setAttribute('title', 'Title');
        wrapper.setAttribute('smartphoneColor', 'smartphoneColor');
        await wrapper.connectedCallback();
        expect(renderLinarChart).toHaveBeenCalledWith({
          data: data.split(','),
          color: 'smartphoneColor',
          element: {
            shadow: wrapper.shadow,
            class: 'lineChart',
            backgroundClass: 'DonutChartComponent__background-title',
          },
        });
      });
    });
  });
  describe('Given method renderTemplate', () => {
    describe('When is called', () => {
      it('Then it should render the correct template', () => {
        wrapper.renderTemplate();
        expect(wrapper.shadow.innerHTML).toMatchSnapshot();
      });
    });
  });
});
