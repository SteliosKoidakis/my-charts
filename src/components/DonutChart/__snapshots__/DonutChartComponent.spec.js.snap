// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given component DonutChartComponent Given method renderTemplate When is called Then it should render the correct template 1`] = `
"
      <div class=\\"DonutChartComponent DonutChartComponent--\\">
        <div class=\\"DonutChartComponent__donut donutChart\\">
          <div class=\\"DonutChartComponent__line-chart lineChart\\">
            <strong class=\\"DonutChartComponent__title\\"></strong>
            <strong class=\\"DonutChartComponent__total\\">undefined</strong>
          </div>
        </div>
        <div class=\\"DonutChartComponent__info\\">
          <div class=\\"DonutChartComponent__tablet-info\\">
            <strong class=\\"DonutChartComponent__tablet-text\\">Tablet</strong>
            <span>0%</span>
            <span class=\\"DonutChartComponent__value\\">undefined</span>
          </div>
          <div class=\\"DonutChartComponent__smartphone-info\\">
            <strong class=\\"DonutChartComponent__smartphone-text\\">Smartphone</strong>
            <span>0%</span>
            <span class=\\"DonutChartComponent__value\\">undefined</span>
          </div>
        </div>
      </div>
    "
`;
